arangod/Cluster/DBServerAgencySync.cpp

```diff
@@ -52,125 +60,176 @@ void DBServerAgencySync::work() {
       << "DBServerAgencySync::execute no vocbase";
     return result;
   }
   
+  Result tmp;
+  VPackBuilder rb;
   auto clusterInfo = ClusterInfo::instance();
   auto plan = clusterInfo->getPlan();
-  auto current = clusterInfo->getCurrent();
-  DatabaseGuard guard(*vocbase);
-  double startTime = TRI_microtime();
-  V8Context* context = V8DealerFeature::DEALER->enterContext(vocbase, true, V8DealerFeature::ANY_CONTEXT_OR_PRIORITY);
+  auto serverId = arangodb::ServerState::instance()->getId();
 
-  if (context == nullptr) {
-    LOG_TOPIC(WARN, arangodb::Logger::HEARTBEAT) << "DBServerAgencySync::execute: no V8 context";
+  VPackBuilder local;
+  Result glc = getLocalCollections(local);
+  if (!glc.ok()) {
     return result;
   }
 
-  TRI_DEFER(V8DealerFeature::DEALER->exitContext(context));
-
-  double now = TRI_microtime();
-
-  if (now - startTime > 5.0) {
-    LOG_TOPIC(WARN, arangodb::Logger::HEARTBEAT) << "DBServerAgencySync::execute took " << Logger::FIXED(now - startTime) << " to get free V8 context, starting handlePlanChange now";
-  }
-
-  auto isolate = context->_isolate;
-
+  auto start = clock::now();
   try {
-    v8::HandleScope scope(isolate);
-
-    // execute script inside the context
-    auto file = TRI_V8_ASCII_STRING(isolate, "handlePlanChange");
-    auto content =
-        TRI_V8_ASCII_STRING(isolate, "require('@arangodb/cluster').handlePlanChange");
-    
-    v8::TryCatch tryCatch;
-    v8::Handle<v8::Value> handlePlanChange = TRI_ExecuteJavaScriptString(
-        isolate, isolate->GetCurrentContext(), content, file, false);
+    // in previous life handlePlanChange
     
-    if (tryCatch.HasCaught()) {
-      TRI_LogV8Exception(isolate, &tryCatch);
-      return result;
-    }
+    VPackObjectBuilder o(&rb);
 
-    if (!handlePlanChange->IsFunction()) {
-      LOG_TOPIC(ERR, Logger::HEARTBEAT) << "handlePlanChange is not a function";
-      return result;
-    }
+    LOG_TOPIC(DEBUG, Logger::MAINTENANCE) << "DBServerAgencySync::phaseOne";
+    tmp = arangodb::maintenance::phaseOne(
+      plan->slice(), local.slice(), serverId, *mfeature, rb);
+    LOG_TOPIC(DEBUG, Logger::MAINTENANCE) << "DBServerAgencySync::phaseOne done";
 
-    v8::Handle<v8::Function> func =
-      v8::Handle<v8::Function>::Cast(handlePlanChange);
-    v8::Handle<v8::Value> args[2];
-    // Keep the shared_ptr to the builder while we run TRI_VPackToV8 on the
-    // slice(), just to be on the safe side:
-    auto builder = clusterInfo->getPlan();
-    args[0] = TRI_VPackToV8(isolate, builder->slice());
-    builder = clusterInfo->getCurrent();
-    args[1] = TRI_VPackToV8(isolate, builder->slice());
-    
-    v8::Handle<v8::Value> res =
-      func->Call(isolate->GetCurrentContext()->Global(), 2, args);
-    
-    if (tryCatch.HasCaught()) {
-      TRI_LogV8Exception(isolate, &tryCatch);
+    LOG_TOPIC(DEBUG, Logger::MAINTENANCE) << "DBServerAgencySync::phaseTwo";
+    glc = getLocalCollections(local);
```

Comment:
Is this redundant to line 140, or is a new copy required after phaseOne completes?  If new copy needed, are there any edge cases of things changing that are not expected to change?

Reply:
This is intentional, phases 1 and 2 are sufficiently independent that it is OK to use a new Local in phase 2. The idea here is that in the meantime some actions might actually have been completed during phase 1 already. I added a comment.